### ISSUE TITLE
API backed User Registration

### ISSUE THREAD

**Issue by jhuntoo** [coreos/dex#255]
I'm implementing user registration in an angular web client, and I'm looking for a method to create a new user, which accepts an email/password

If i'm not mistaken there are 2 ways to create a user
1. the registration flow using the server generated pages
2. the admin api

Option 2  is not suitable as it requires admin privileges. Option 1 isn't ideal for a client side web app.

Am I missing something here ? 

Could we allow the POST users endpoint to be configured to allow anonymous access? Or create a new endpoint for that purpose ?

**Comment by ericchiang** [comment]
See [this comment](https://github.com/coreos/dex/issues/225#issuecomment-165535788).

> This sounds like the OIDC [implicit flow](http://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth) (though I admittedly have very little experience with native client apps):

> > The Implicit Flow is mainly used by Clients implemented in a browser using a scripting language. The Access Token and ID Token are returned directly to the Client, which may expose them to the End-User and applications that have access to the End-User's User Agent. The Authorization Server does not perform Client Authentication.
I don't think dex currently supports this. It's not that its design prevents implicit flow support, but CoreOS doesn't use that flow internally so it's not been tested.

> It's not on the immediate roadmap, though we'd love to see dex support this.

The correct way to support applications like this would be to implement the OIDC implicit flow for browser app authentication.

**Comment by jhuntoo** [comment]
@ericchiang that looks fine for authenticating a user via a browser app. However what I'm stuck on at the moment is how to create a local users in the first place.

I was expecting to find an endpoint like `/users` to which I could POST a JSON request like
`
{
   "email" : "joe@example.com"
   "password" : "secret"
}`

the User API allows something like this. See the request struct [here](https://github.com/coreos/dex/blob/7c8e25f2923e71b9736c14abdf25c9acb69c9984/schema/workerschema/v1-gen.go#L122-L126), however this limited to admin users only because of the check performed [here](https://github.com/coreos/dex/blob/7c8e25f2923e71b9736c14abdf25c9acb69c9984/user/api/api.go#L142).

Creating users via a browser app, would need to be publicly available to non authenticated, non admin requests.
 
The existing user api endpoint does not seem suitable for this use case, as it allows you to set sensitive properties such as the admin flag, which should not be exposed.

```
type User struct {
	Admin bool 'json:"admin,omitempty"'
	CreatedAt string 'json:"createdAt,omitempty"'
	Disabled bool 'json:"disabled,omitempty"'
	DisplayName string 'json:"displayName,omitempty"'
	Email string 'json:"email,omitempty"'
	EmailVerified bool 'json:"emailVerified,omitempty"'
	Id string 'json:"id,omitempty"'
}
```

What I could suggest is a new `/registrations` endpoint which permits the request I detailed above, which would create the new user, as non admin by default, which the supplied email/password, and return a response which indicates whether it succeeded or not.

Does that seem reasonable to you ? If so I'm happy to submit a PR for this functionality.

**Comment by ericchiang** [comment]
> What I could suggest is a new /registrations endpoint which permits the request I detailed above, which would create the new user, as non admin by default, which the supplied email/password, and return a response which indicates whether it succeeded or not.

dex already has a registration through its web UI. Do an OIDC redirect to dex's login page, the user registers, then they're redirected back to your app with an identity. Does that existing workflow not work for this scenario, and if so how is this proposed solution different?

**Comment by jhuntoo** [comment]
The existing registration flow only allows the presentation to be defined by the [html / go template](https://github.com/coreos/dex/blob/7c8e25f2923e71b9736c14abdf25c9acb69c9984/static/html/register.html), and be rendered server-side. This removes the ability to put the presentation logic, in the front-end application where, IMO, it belongs.

The solution I propose is different in that it would allow clients, in particular client-side web apps, to complete control the UX around signup / registration, choice of presentation tech. Maintaining a web app that is split across a front-end codebase with a little bit hidden away in dex isn't ideal, quickly frankly it's messy. Dex as an identity management system, shouldn't impose itself on front-end decisions.

@ericchiang  Would a more detailed proposal help to explain what I'd like to achieve with the new API endpoint?

**Comment by bobbyrullo** [comment]
The idea of having login and registration UI live within dex is not accidental - we wanted a way for people to use dex as a completely independent component which provides registration and OIDC authentication out of the box, without additional engineering. Our opinionated stance is that endlessly re-inventing authentication flows with every app is dangerous anti-pattern. Not to mention tedious.

However, we don't want dex to impose itself any more than necessary, which is why, apart from login and registration, there's no UI, just APIs.

For your case, the user [management API](https://github.com/coreos/dex/blob/master/schema/workerschema/v1.json) might be able to provide what you need. It provides a set of (right now, rudimentary) methods for creating and manipulating users. It *is* restricted to admin users....and admin clients. So the idea here is that you'd create an endpoint on your app (which is registered as an admin client...more on that later) and your app would make the call to dex with its credentials.

There's two challenges here right now:

* It still requires a privileged user. In your case you could create an admin user, put it's credentials in a secret somewhere, and use that as the user which accesses the API

* The other challenge is that there's no way to flip the client admin bit except by poking the DB which is pretty gross. We can fix that with `dexctl` ASAP if this is a priority for you. 

I'm not completely against a new endpoint but I do have a couple of reservations:
* I don't love the idea of Yet Another User Way Users are Created. Just widens the attack surface, and is another thing that would need to be locked down in case of a DOS
*  The registration endpoint you sketch out only works for the "local" connector - how would you handle registration with Google, Github, etc.?

**Comment by bobbyrullo** [comment]
Also some background on why the user management API is what it is:

We wanted an API for managing users. We needed an UI, but we didn't want to add UI to dex itself, so we decided to let it be in a client. And we wanted only admin users to perform these tasks, so we require that the client pass an admin user's credentials. And, since an admin user might authenticate to an untrusted client, we require that the app be blessed as well.

**Comment by jnflint** [comment]
Adding to this issue as it is relevant to my case. For my use case I will probably use the user management API. I am collecting more information such as DOB, phone and address plus an approval process which might be from credit card processing call back or a manual approval. I had thought to use the existing self registration and then gather the additional information as the next step after the user is authenticated but that method is fixed to requesting emailaddress and full name whereas I need the name parts (first, surname etc and names are hard to parse). Since the registration is not suitable it means I can't use the login either as they are tied together. So I prefer the idea of API only with any examples being a reference implementation and not tied in.

On the user management, is there any thought to implementing the standard user profile claims (http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) as user attributes and the possibility of extending the user for additional attributes or should this be done outside this system?

**Comment by jhuntoo** [comment]
I've been pondering this issue for the last 2 days. The out-of-box registration is fine, and is great for those who want something safe and rock-solid, though in my particular case I'm soon forced to make compromises with UX.

While this change would be a slight departure from dex's original mandate of "taking care of it all", with a well reviewed implementation and guidance to implementing an API backed registration/login flow, dex's usefulness could have a far broader appeal and could still help implementers fall into the pit of success.

User registration is one of those areas in UI's where many like control. e.g. UX is super important when trying to increase conversion rates.

@bobbyrullo Would you entertain a more detailed proposal which address your concerns listed above?

**Comment by bobbyrullo** [comment]
@jhuntoo - Sure, I would love to hear your ideas in this area. A couple of more thoughts that guided some of the design of dex: We had envisioned (and still do!) dex as an Identity provider that, like the way people use Google OAuth and OIDC for authentication and identity,  provides identity for a number of applications (clients). 

In such cases people don't expect to customize the Google registration experience, and instead ask them to authenticate via Google (or Facebook or whatever) and then add then perhaps they direct the user to fill in additional information specific to their app. So that's how I expected people to use dex.

However, I recognize that there are differences - for one, you can't rely on people having an account on your dex installation like you can for Google. And also, it seems that many folks (including yourself perhaps) are looking to use dex as the identity solution for a single app, and want to streamline the process so there's as few steps for the user as possible. It would be great if we could support that as well.

Some things I would like to see in your proposal considered are:
* support for remote identities 
* preventing abuse - make it possible to prevent or make difficult people from creating tons of accounts 
* ease of integration

Thanks again.

**Comment by bobbyrullo** [comment]
@jnflint - are you looking store all your app-specific stuff (CC info etc.) in dex as well?

Also just because you're not using the registration doesn't mean you can't use the login - you can turn off registration with a flag and just use your the user management API for custom "registration"

**Comment by bobbyrullo** [comment]
@jhuntoo: Also, I think you know this, but you point to wherever you like for the HTML templates. I know this forces you to use Go templates, but it does give you a fair amount of flexibility.

**Comment by jnflint** [comment]
@bobbyrullo good to know I can turn the registration off. I was thinking to store profile/identity related attributes. DOB, contact info etc. (definitely not cc info) plus group membership when that is supported. My use is for a membership database as well as using dex for SSO for a variety of apps. We have a PHP based CMS, but rather than building every future application into that we will integrate the identity management / authentication SSO system into future app which might be based on go, ruby, java or whatever. The membership system is pretty much only user profile info plus membership renewal dates, so I thought I can shoehorn a few extra fields in. It might be that part of the application grows so it may be better I separate the membership stuff out so it can be extended without having to worry about keeping everything in sync.

**Comment by bobbyrullo** [comment]
closing this out for now, as it doesn't really mesh with current vision of dex (i.e. dex is not a general user data store)
